File: rust/crates/dirlens-core/src/analysis/todo.rs
//! TODO/FIXME 抽出（-K）。dirlens.py の scan_todos の等価移植。

use std::sync::OnceLock;

use regex::Regex;

use crate::pyc::{char_len, py_strip, truncate_chars};

fn todo_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"(?i)\b(TODO|FIXME|HACK|XXX)\b").unwrap())
}

/// 戻り値: (行番号, 種別（大文字化）, スニペット)
pub fn scan_todos(text: &str) -> Vec<(usize, String, String)> {
    if text.is_empty() {
        return Vec::new();
    }
    let mut results = Vec::new();
    for (i, line) in text.split('\n').enumerate() {
        if let Some(m) = todo_re().captures(line) {
            let kind = m.get(1).unwrap().as_str().to_uppercase();
            let mut snippet = py_strip(line).to_string();
            if char_len(&snippet) > 80 {
                let (head, _) = truncate_chars(&snippet, 80);
                snippet = format!("{}…", head);
            }
            results.push((i + 1, kind, snippet));
        }
    }
    results
}

File: swift/Sources/DirlensCore/Analysis/Todo.swift
// TODO/FIXME 抽出（-K）。rust/crates/dirlens-core/src/analysis/todo.rs の等価移植。

private let todoRe = Regexp("\\b(TODO|FIXME|HACK|XXX)\\b", caseInsensitive: true)

/// 戻り値: (行番号, 種別（大文字化）, スニペット)
public func scanTodos(_ text: String) -> [(Int, String, String)] {
    if text.isEmpty { return [] }
    var results: [(Int, String, String)] = []
    for (i, line) in splitLines(text).enumerated() {
        let lineStr = line
        if let m = todoRe.firstMatch(lineStr), let kind = m[1] {
            var snippet = pyStrip(lineStr)
            if charLen(snippet) > 80 {
                let (head, _) = truncateChars(snippet, 80)
                snippet = head + "…"
            }
            results.append((i + 1, kind.uppercased(), snippet))
        }
    }
    return results
}
